lib/bb/runqueue.py

```diff
@@ -0,0 +1,3116 @@
+
+        self.state = runQueuePrepare
+
+        # For disk space monitor
+        # Invoked at regular time intervals via the bitbake heartbeat event
+        # while the build is running. We generate a unique name for the handler
+        # here, just in case that there ever is more than one RunQueue instance,
+        # start the handler when reaching runQueueSceneInit, and stop it when
+        # done with the build.
+        self.dm = monitordisk.diskMonitor(cfgData)
+        self.dm_event_handler_name = '_bb_diskmonitor_' + str(id(self))
+        self.dm_event_handler_registered = False
+        self.rqexe = None
+        self.worker = {}
+        self.fakeworker = {}
+
+    def _start_worker(self, mc, fakeroot = False, rqexec = None):
+        logger.debug("Starting bitbake-worker")
+        magic = "decafbad"
+        if self.cooker.configuration.profile:
+            magic = "decafbadbad"
+        fakerootlogs = None
+        if fakeroot:
+            magic = magic + "beef"
+            mcdata = self.cooker.databuilder.mcdata[mc]
+            fakerootcmd = shlex.split(mcdata.getVar("FAKEROOTCMD"))
+            fakerootenv = (mcdata.getVar("FAKEROOTBASEENV") or "").split()
+            env = os.environ.copy()
+            for key, value in (var.split('=') for var in fakerootenv):
+                env[key] = value
+            worker = subprocess.Popen(fakerootcmd + ["bitbake-worker", magic], stdout=subprocess.PIPE, stdin=subprocess.PIPE, env=env)
+            fakerootlogs = self.rqdata.dataCaches[mc].fakerootlogs
+        else:
+            worker = subprocess.Popen(["bitbake-worker", magic], stdout=subprocess.PIPE, stdin=subprocess.PIPE)
+        bb.utils.nonblockingfd(worker.stdout)
+        workerpipe = runQueuePipe(worker.stdout, None, self.cfgData, self, rqexec, fakerootlogs=fakerootlogs)
+
+        workerdata = {
+            "taskdeps" : self.rqdata.dataCaches[mc].task_deps,
+            "fakerootenv" : self.rqdata.dataCaches[mc].fakerootenv,
+            "fakerootdirs" : self.rqdata.dataCaches[mc].fakerootdirs,
+            "fakerootnoenv" : self.rqdata.dataCaches[mc].fakerootnoenv,
+            "sigdata" : bb.parse.siggen.get_taskdata(),
+            "logdefaultlevel" : bb.msg.loggerDefaultLogLevel,
+            "build_verbose_shell" : self.cooker.configuration.build_verbose_shell,
+            "build_verbose_stdout" : self.cooker.configuration.build_verbose_stdout,
+            "logdefaultdomain" : bb.msg.loggerDefaultDomains,
+            "prhost" : self.cooker.prhost,
+            "buildname" : self.cfgData.getVar("BUILDNAME"),
+            "date" : self.cfgData.getVar("DATE"),
+            "time" : self.cfgData.getVar("TIME"),
+            "hashservaddr" : self.cooker.hashservaddr,
+            "umask" : self.cfgData.getVar("BB_DEFAULT_UMASK"),
+        }
+
+        worker.stdin.write(b"<cookerconfig>" + pickle.dumps(self.cooker.configuration) + b"</cookerconfig>")
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The use of `subprocess.Popen.stdout.write()` might deadlock if any OS pipe buffers fill up and block the child process. We recommend using the `communicate()` API.

[Learn more](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate)

Similar issue at line numbers 1291 and 1292.